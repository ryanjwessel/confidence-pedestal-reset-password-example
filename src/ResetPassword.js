import React from "react";

const initialState = {
  currentPassword: "",
  newPassword: "",
  retypePassword: "",
  isSubmitted: false,
  isValid: false
};

const validatePasswords = (currentPassword, newPassword, retypePassword) => {
  return (
    currentPassword.length > 0 &&
    newPassword.length > 0 &&
    retypePassword.length > 0 &&
    newPassword === retypePassword
  );
};

const ResetPassword = () => {
  const [formState, setFormState] = React.useState(initialState);
  const {
    currentPassword,
    newPassword,
    retypePassword,
    isSubmitted,
    isValid
  } = formState;

  // Validate the password fields anytime the input fields are updated.
  React.useEffect(() => {
    setFormState(prevFormState => ({
      ...prevFormState,
      isValid: validatePasswords(currentPassword, newPassword, retypePassword)
    }));
  }, [currentPassword, newPassword, retypePassword]);

  const handleChange = ({ target: { name, value } }) => {
    setFormState({
      ...formState,
      [name]: value
    });
  };

  const handleSubmit = () => {
    setFormState({
      ...formState,
      isSubmitted: true
    });
  };

  return (
    <>
      <div>
        <label id="currentPasswordLabel" htmlFor="currentPassword">
          Current Password
        </label>
        <input
          type="password"
          name="currentPassword"
          value={currentPassword}
          onChange={handleChange}
          aria-labelledby="currentPasswordLabel"
        />
      </div>
      <div>
        <label id="newPasswordLabel" htmlFor="newPassword">
          New Password
        </label>
        <input
          type="password"
          name="newPassword"
          value={newPassword}
          onChange={handleChange}
          aria-labelledby="newPasswordLabel"
        />
      </div>
      <div>
        <label id="retypePasswordLabel" htmlFor="retypePassword">
          Retype Password
        </label>
        <input
          type="password"
          name="retypePassword"
          value={retypePassword}
          onChange={handleChange}
          aria-labelledby="retypePasswordLabel"
        />
      </div>
      <button type="submit" disabled={!isValid} onClick={handleSubmit}>
        Submit
      </button>
    </>
  );
};

export default ResetPassword;
