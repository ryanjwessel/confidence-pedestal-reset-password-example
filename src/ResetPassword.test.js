import React from "react";
import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import ResetPassword from "./ResetPassword";

describe("ResetPassword", () => {
  const mockCurrentPassword = "oldPassword9";
  const mockNewPassword = "newPassword10@";
  const mockNonMatchingNewPassword = "newPassword11&";

  it("should match the snapshot", () => {
    const { container } = render(<ResetPassword />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("the submit button should only be enabled when the form is valid", () => {
    render(<ResetPassword />);

    const currentPassword = screen.getByLabelText("Current Password");
    const newPassword = screen.getByLabelText("New Password");
    const retypePassword = screen.getByLabelText("Retype Password");
    const submitButton = screen.getByRole("button");

    expect(submitButton).toBeDisabled();

    fireEvent.change(currentPassword, {
      target: { value: mockCurrentPassword }
    });
    expect(currentPassword.value).toEqual(mockCurrentPassword);

    fireEvent.change(newPassword, {
      target: { value: mockNewPassword }
    });
    expect(newPassword.value).toEqual(mockNewPassword);

    fireEvent.change(retypePassword, {
      target: { value: mockNewPassword }
    });
    expect(retypePassword.value).toEqual(mockNewPassword);

    // Everything is valid, button should be enabled now.
    expect(submitButton).toBeEnabled();

    fireEvent.change(retypePassword, {
      target: { value: mockNonMatchingNewPassword }
    });
    expect(retypePassword.value).toEqual(mockNonMatchingNewPassword);

    // New passwords are no longer matching. Submit button should be disabled.
    expect(submitButton).toBeDisabled();
  });
});
