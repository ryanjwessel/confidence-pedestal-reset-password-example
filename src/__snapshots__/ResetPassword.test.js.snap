// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword should match the snapshot 1`] = `
<div>
  <label
    for="currentPassword"
    id="currentPasswordLabel"
  >
    Current Password
  </label>
  <input
    aria-labelledby="currentPasswordLabel"
    name="currentPassword"
    type="password"
    value=""
  />
</div>
`;
